Fix shell to compile with c++

nordic-krch opened the pull request:
Here is the first step in direction of shell being cpp friendly. Well, this step is enough but second step would be required to keep zephyr consistent.

There are some cosmetics to allow cpp compilation and one api related change. It's around creation of subcommand array. The syntax we had now cannot be transferred to c++ (see #13257 ).

As i didn't want to break API, i created macro `SHELL_STATIC_SUBCMD_SET_CREATE` (compared to current one `SHELL_CREATE_STATIC_SUBCMD_SET`). The plan is to deprecate old macro (and replace usage across tree) in a separate PR. I moved verb to the end of the macro name to be in line with `SHELL_CMD_REGISTER`. I also added macro `SHELL_DYNAMIC_CMD_CREATE` and plan to deprecate `SHELL_CREATE_DYNAMIC_CMD` to have all naming consistent across the file.

@nashif please let me know if that could go to 1.14. There is no functional changes planned. it's just about structures initialization and this is a bug since code does not compile in c++.

nordic-krch committed:
shell: Refactor shell argument count validation

Shell arguments structure was stored as a pointer in shell structure
and NULL pointer indicated that argument count checking is skipped.
It has been reworked to hold the structure (2 bytes) in the shell
structure with mandatory=0 skipping the check. This approach is cpp
friendly, contrary to the legacy one.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
shell: Add cpp-friendly shell macro for subcommands creation

Add macro for creating subcommand set in a way that is
accepted by C++ (SHELL_STATIC_SUBCMD_SET_CREATE). Currently,
it exists along with SHELL_CREATE_STATIC_SUBCMD_SET which is
used in the tree but it is not liked by cpp.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@rodrigopex, could you try it out?

rodrigopex commented on include/shell/shell.h:
```diff
@@ -137,7 +126,7 @@ struct shell_static_entry {
 			STRINGIFY(UTIL_CAT(shell_root_cmd_, syntax)))))	   \
 	__attribute__((used)) = {					   \
 		.is_dynamic = false,					   \
```
This will not work for C++. You cannot use structure initializer that name the fields, ```.is_dynamic = false,``` will not compile. The better way to keep it running in both C and C++ is to set directly like ```__attribute__((used)) = {					   
		 false, {&UTIL_CAT(_shell_, syntax)}} ``` maybe you'll need a cast to the first union field.

rodrigopex commented on include/shell/shell.h:
```diff
@@ -188,12 +203,21 @@ struct shell_static_entry {
  * @param[in] get	Pointer to the function returning dynamic commands array
  */
-#define SHELL_CREATE_DYNAMIC_CMD(name, get)		\
+#define SHELL_DYNAMIC_CMD_CREATE(name, get)		\
```
Same here.

rodrigopex commented on include/shell/shell.h:
```diff
@@ -210,11 +234,10 @@ struct shell_static_entry {
 #define SHELL_CMD_ARG(_syntax, _subcmd, _help, _handler,		      \
 		      _mandatory, _optional) {				      \
```
Same here.

rodrigopex commented on include/shell/shell.h:
```diff
@@ -159,7 +148,32 @@ struct shell_static_entry {
+		__VA_ARGS__,						\
+		SHELL_SUBCMD_SET_END					\
+	};								\
+	static const struct shell_cmd_entry name = {			\
```
Same here.

rodrigopex commented on subsys/shell/shell.c:
```diff
@@ -524,23 +524,13 @@ static int exec_cmd(const struct shell *shell, size_t argc, char **argv,
-			ret_val = cmd_precheck(shell,
-					       (args->mandatory == argc));
-		}
+	if (shell->ctx->active_cmd.args.mandatory) {
```
Is it possible to have 1 optional and no mandatory?

rodrigopex commented:
> @rodrigopex, could you try it out?

I have put some comments there? Didn't you change the shell_cmds.c? I was thinking it was necessary based on the fact the command name and usage changed. I did that here: https://github.com/zephyrproject-rtos/zephyr/compare/master...rodrigopex:i13257

rodrigopex reviewed (commented):
~This code will not compile in C++~. I am thinking about the struct initializer... Did somebody add some different flag to the g++ compiler? Code like ```struct foo a = {f. = 1, .x = 2};``` didn't compile today morning...

rodrigopex reviewed (commented):
The code is running ok! :+1:

dbkinder reviewed (approved):
+1 for doc changes

jarz-nordic commented on include/shell/shell.h:
```diff
@@ -159,7 +148,32 @@ struct shell_static_entry {
+		__VA_ARGS__,						\
+		SHELL_SUBCMD_SET_END					\
+	};								\
+	static const struct shell_cmd_entry name = {			\
```
Do we consider removing `static` here? Variable name could be more complex like: `shell_root_cmd_##name`

nordic-krch replied:
right, i think if i add prefix we could remove static.

nordic-krch committed:
shell: Modify subcommands to use SHELL_STATIC_SUBCMD_SET_CREATE

It is planned to deprecate SHELL_CREATE_STATIC_SUBCMD_SET macro
which is replaced by SHELL_STATIC_SUBCMD_SET_CREATE.

Additionally, removed irrelevant comments about alphabetical
ordering which is no longer needed.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
shell: Deprecate macros for subcommands creation

Macros are replaced by C++ friendly versions:
- SHELL_CREATE_STATIC_SUBCMD_SET by SHELL_STATIC_SUBCMD_SET_CREATE
- SHELL_CREATE_DYNAMIC_CMD by SHELL_DYNAMIC_CMD_CREATE

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
shell: Making root command a global variable

By making it global we ensure that there will not be the
same root commands.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

carlescufi commented:
@rodrigopex @jarz-nordic can you revisit this please?

jarz-nordic commented:
@carlescufi : why shippable is reporting not finished task?

jakub-uC reviewed (approved):
Looks good.

jarz-nordic commented on doc/reference/shell/index.rst:
```diff
@@ -109,13 +107,11 @@ subcommands.
-		SHELL_SUBCMD_SET_END /* Array terminated. */
-	};
+		SHELL_CMD(ping,   NULL, "Ping command.", cmd_demo_ping)
+	);
```
Missing: `SHELL_SUBCMD_SET_END `

nordic-krch replied:
fixed

jakub-uC reviewed (changes requested):
After updating concept: bringing back `SHELL_SUBCMD_SET_END`, you have forgotten to update documentation
